### PR TITLE
Draft: Terminal SpanKind

### PR DESCRIPTION
**Pull request by CodeBlanch**
The goal is to have some kind of internal span that isn't collected by default. For example, Http POSTs to Zipkin. There was a comment in `HttpClientCollectorOptions` calling these "terminal spans" so I ran with it.

This PR is a potential design. Trying to gather some feedback before opening a PR into spec.

* Added SpanKind.Terminal. That's the spec bit.
* Updated SpanSdk so that IsRecording = false when SpanKind = Terminal. Also, if parent has IsRecording=false, so too will children. Fruit of the poison tree, so to speak.
* Zipkin and Jaeger updated to start Terminal spans in their export work. In order to do that I needed a reference to the TracerFactory.Default which didn't seem to be set ever, so I updated the factory Create method to do that. Not sure if that was a bug/oversight or done on purpose? I just noticed there is also #630 for this issue.
* AppInsights uses TelemetryClient, which itself acts as a kind of exporter (buffering, etc). Not quite sure how to get it to start a Terminal span when it actually goes to transmit data. Might need to have our own sink that is set on AppInsights TelmetryConfiguration?

Note: Haven't tried to run tests on this work yet. Updates are probably needed for the filter changes.

cc @SergeyKanzhelev @pakrym

### CODE REVIEW

**Comment by SergeyKanzhelev** [comment]
I'll review the proposed a bit later. While I remember - here is the link from @lmolkova on her thoughts for inspiration: https://github.com/lmolkova/oteps/blob/terminal-context/text/0069-terminal-context.md

**Comment by CodeBlanch** [comment]
@SergeyKanzhelev Thanks for the link. I'm open to whatever.

Mine on here is basically...
```
var span = StartSpan(SpanKind.Terminal);
try { ...export... }
finally { EndSpan(span); }
```
[If I'm reading it correctly] @lmolkova's is basically...
```
DistributedContext.Current.CollectionMode = CollectionMode.OFF;
try { ...export... }
finally { DistributedContext.Current.CollectionMode = CollectionMode.ON; }
```

Pretty similar, eh?

Neither really takes care of the AppInsights thing. In the case of TelemetryClient the actual transmission can fire on another thread at a later time so the span will have likely already ended in the first case, or the CollectionMode will have already switched back for DistributedContext case, or, worse, some random Async state will flow to the thread that does the transmission and it will be a race what mode it ends up with.

**Comment by cijothomas** [comment]
Correct me if misunderstood the issue.
This issue is not specific to OT-dotnet right? Why don't we post the proposal in the https://github.com/open-telemetry/oteps repo, instead of having a Dotnet only solution?

**Comment by cijothomas** [comment]
> Correct me if misunderstood the issue.
> This issue is not specific to OT-dotnet right? Why don't we post the proposal in the https://github.com/open-telemetry/oteps repo, instead of having a Dotnet only solution?

Discard this comment! I didnt fully read your description!

File: src/OpenTelemetry.Exporter.Zipkin/ZipkinTraceExporter.cs [hunk 1]
```diff
@@ -73,6 +76,13 @@ public override Task ShutdownAsync(CancellationToken cancellationToken)
 
         private Task SendSpansAsync(IEnumerable<SpanData> spans)
         {
+            if (this.tracer == null)
+            {
+                this.tracer = TracerFactoryBase.Default.GetTracer("OpenTelemetry");
+            }
+
+            using var scope = this.tracer.StartActiveSpan("OpenTelemetry.Zipkin.SendSpans", SpanKind.Terminal, out _);
```
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Zipkin/ZipkinTraceExporter.cs:84]: The issue you described about application insights telemetryclient doing its own  bacthing and making the http.send at some future point from different thread - this issue is not specific to application insights right? Any library can face this?

Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Zipkin/ZipkinTraceExporter.cs:84]: @cijothomas Correct. It becomes the responsibility of the exporter lib author to properly flag the span as terminal (or set the CollectionMode=OFF) when it goes to transmit data. It's not ideal.

Comment (cijothomas) [reply src/OpenTelemetry.Exporter.Zipkin/ZipkinTraceExporter.cs:84]: mm. Got it.

**Comment by SergeyKanzhelev** [comment]
Sorry took me a while. As there is no ideal solution, here are some thinking and recommendations:

Some thoughts of `SpanKind.Terminal`:

- Can span be `SpanKind.Producer` and Terminal at the same time? Let's say there is a queue library that uses some common RPC protocol to queue a message. And the library desires to NOT collect the Span representing that RPC call. It must mark the Span produced as both - `Terminal` and `Producer`.
- The requirement to start a span to mark the current thread as `Terminal` is not intuitive for the scenario of implementing an exporter. 
- The good thing about `SpanKind.Terminal` is that the same `AsyncLocal` variable is used which minimizes an overhead - no need to read an additional `AyncLocal` on every `Span` start.

Perhaps we can combine these two approaches - make an additional variable on Span that will indicate whether children must be collected.

@tarekgh what do you think? On one hand we want to minimize the API surface on `Activity/Span`. On other hand, introducing a new API that indicates `CollectionMode` may be an overkill. Please note, it must be part of API as @pakrym and other library authors will need to mark Spans this way.

**Comment by tarekgh** [comment]
I have a question just to make sure I understand the whole issue, Do we always consider when the `Span.IsRecording==false`  means the Span is in terminal state? or we have cases `Span.IsRecording=false` and the Span shouldn't be treated as terminal?

**Comment by CodeBlanch** [comment]
@tarekgh If you are using a sampler it could set IsRecording = false if it decides it isn't going to sample a particular span.

**Comment by pakrym** [comment]
Can be this accomplished by having the ability to customize what spans customer subscribes to and having some rules that exclude built in spans by default. 

It seems that decision on if customer wants to see a span in his trace is on customers.

**Comment by CodeBlanch** [comment]
@pakrym It would be easy to do once we have identified them, somehow. Today it's very hard to do. We can't tell internal stuff from external. This is what we do today (URL sniff):

https://github.com/open-telemetry/opentelemetry-dotnet/blob/93fa3f9ff5f3275b6924746f9091b0eda9b931dd/src/OpenTelemetry.Adapter.Dependencies/HttpClientAdapterOptions.cs#L73-L83

If we can identify them all as special, then it would be easy to have an option or some kind of mechanism for the user to decide if they want them included in their data or not. BUT that in itself is tricky. If you decide you want, for example, the call to Zipkin to be reported, the call itself generates a new span which generates a new call which generates a new span which generates a new call on and on forever.

**Comment by SergeyKanzhelev** [comment]
> It seems that decision on if customer wants to see a span in his trace is on customers not library authors.

@pakrym another aspect is for an SDK that wants to hide an implementation detail of an underlying call. For instance Azure SDK instrumented to report Span, while not reporting an underlying http call, collected by the automatic collection. Same for EF and SQL.

@tarekgh , it's ultimately an `IsRecording` flag. But the library author or exporter's author wants to inform that all "future" spans must not be recorded.

**Comment by tarekgh** [comment]
> it's ultimately an IsRecording flag. But the library author or exporter's author wants to inform that all "future" spans must not be recorded.

Could you please clarify more regarding the statement `all "future" spans must not be recorded`? does this mean the future spans will be automatically marked with the Terminal kind? or the span collectors will see the first terminal span and then stop recording any future spans regardless of its state? I am just trying to understand the end-to-end scenario and how the Terminal really gets used.

**Comment by SergeyKanzhelev** [comment]
@tarekgh terminal span or `CollectionMode.OFF` means that all children of this span are likely not interesting. Th default behavior of SDK would be to not collect information about those spans. Basically, it will be sampler functionality to check on every span creation whether the current thread is currently marked as terminal and set `IsRecording=false` for children.

I think the property of being terminal must be set explicitly by the user. Collection logic (SDK) will react on this property the way it's configured.

**Comment by tarekgh** [comment]
@SergeyKanzhelev thanks. I think the options to support that (which similar to what proposed) are:

- Introduce CollectionMode (or we can call it more generic mode that works with the collection scenario and other scenarios if there is any). Currently, only Termination mode is required. do we think we can have more scenarios that need to add more values to the mode? if yes, then CollectionMode will be a reasonable option to go with. Otherwise, use SpanKind instead to include it.
- Add Terminal to SpanKind as proposed. This option will be reasonable if we think we need to support the Terminal scenario only and unlikely we need to add more like that in the future. If we think Producer can be combined with Terminal, then we can have extra value like TerminalProducer. But if we think we are going to combine more values with Terminal, then I would say to go to with the first option having CollectionMode or we'll make SpanKind as bit flags and handle the validation of the right combination.


CC @noahfalk

**Comment by noahfalk** [comment]
> The good thing about SpanKind.Terminal is that the same AsyncLocal variable is used which minimizes an overhead - no need to read an additional AyncLocal on every Span start.

I think we would need to benchmark before making that argument. Educated guess - creating the extra Activity will cost us far more time than any savings from eliminating an AsyncLocal read resulting in overall perf loss.

> it must be part of API as @pakrym and other library authors will need to mark Spans this way.

I think there are two different scenarios here that are similar, but they aren't necessarily going to have the same solution. 
 - The scenario that instigated this issue (as I understand it) is a desire for exporters to not accidentally log recursively. On its own this doesn't automatically require a public API because exporters can communicate with themselves via private thread-local or async-local
 - A broader scenario is that libraries that are not the exporter may want a way to communicate with the exporter to influence what telemetry is captured. This one probably does need some public API or agreed upon convention.

Where possible it is always nice to have a generalized solution that handles as many cases as possible, but my impression is that the 2nd scenario there is going to require considerable design effort and a far more varied set of requirements. We are probably far safer designing a solution to the first scenario that does not introduce any new API so that we are immediately unblocked. If a suitably generalized solution to the 2nd scenario materializes it should be easy to switch from a private mechanism to a general purpose mechanism. I see that Liudmilla's proposal was December last year but not sure if there is any ongoing discussion related to it and whether it has achieved broader acceptance at this point?

In short, I don't think we should be adding new public API just yet : ) I'd be happy to discuss Liudmilla's proposal more if it is a priority but I can't tell if it was and it seems much broader scope than the current issue here.

**Comment by reyang** [comment]
In Python a separate concept [suppress_instrumentation](https://github.com/open-telemetry/opentelemetry-python/blob/8d09319c43a24b05d14128361de2c9afe8c856b6/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py#L78) is used.

We need to consider this across logs/metrics/traces - if there is a metrics exporter which also collects metrics, `SpanKind` probably won't help us.

**Comment by CodeBlanch** [comment]
@reyang I like the idea of having something that works potentially for traces, metrics, & logs.
